skbio/metadata/_interval.py

```diff
@@ -0,0 +1,699 @@
+
+    >>> im.sort()
+    >>> im    # doctest: +ELLIPSIS
+    3 interval features
+    -------------------
+    Interval(interval_metadata=..., locations=[(1, 2), (4, 7)], \
+boundaries=[(True, True), (True, True)], metadata={'gene': 'sagA'})
+    Interval(interval_metadata=..., locations=[(3, 7)], \
+boundaries=[(True, True)], metadata={'gene': 'sagC'})
+    Interval(interval_metadata=..., locations=[(3, 9)], \
+boundaries=[(True, True)], metadata={'gene': 'sagB'})
+
+    Query the genes by location and/or metadata:
+
+    >>> q1 = im.query([(7, 9)])
+    >>> list(q1)  # doctest: +ELLIPSIS
+    [Interval(interval_metadata=..., locations=[(3, 9)], \
+boundaries=[(True, True)], metadata={'gene': 'sagB'})]
+    >>> q2 = im.query(metadata={'gene': 'sagA'})
+    >>> list(q2)  # doctest: +ELLIPSIS
+    [Interval(interval_metadata=..., locations=[(1, 2), (4, 7)], \
+boundaries=[(True, True), (True, True)], metadata={'gene': 'sagA'})]
+    >>> q3 = im.query([(1, 2)], metadata={'gene': 'foo'})
+    >>> list(q3)
+    []
+
+    Drop genes:
+
+    >>> im.drop(metadata={'gene': 'sagA'})
+    >>> im.sort()
+    >>> im   # doctest: +ELLIPSIS
+    2 interval features
+    -------------------
+    Interval(interval_metadata=..., locations=[(3, 7)], \
+boundaries=[(True, True)], metadata={'gene': 'sagC'})
+    Interval(interval_metadata=..., locations=[(3, 9)], \
+boundaries=[(True, True)], metadata={'gene': 'sagB'})
+    """
+    def __init__(self):
+        # List of Interval objects.
+        self._intervals = []
+
+        # IntervalTree object to allow faster querying of interval objects.
+        self._interval_tree = IntervalTree()
+
+        # Indicates if the IntervalTree needs to be rebuilt.
+        self._is_stale_tree = False
+
+    def _rebuild_tree(method):
+        """Rebuild the IntervalTree."""
+        @wraps(method)
+        def inner(self, *args, **kwargs):
+            if self._is_stale_tree is False:
+                return method(self, *args, **kwargs)
+            self._interval_tree = IntervalTree()
+            for f in self._intervals:
+                for start, end in f.locations:
+                    self._interval_tree.add(start, end, f)
+            self._is_stale_tree = False
+            return method(self, *args, **kwargs)
+        return inner
+
+    def _reverse(self, length):
+        """Reverse the location coordinates with given length.
+
+        This modifies the coordinates by subtracting them from the
+        given length. This operation facilitates the sequence
+        reverse complement.
+
+        Parameters
+        ----------
+        length : int
+            Largest end coordinate to perform reverse complement.
+            This typically corresponds to the length of sequence.
+        """
+        for f in self._intervals:
+            # staled doesn't need to be called, since the setter for
+            # Interval will take care of this
+            intvls = list(map(lambda x: (length-x[1], length-x[0]),
```

Comment:
Use a list comprehension instead of `map` and `list` cast. Python 3 discourages use of `map`